### PR TITLE
Redact credentials written as JSON properties (#1497)

### LINKED ISSUE #1497: SensitiveErrorTextRedactor: redact credentials written as JSON properties ("password": "…"), incl. \u0022-escaped form
## Summary

`SensitiveErrorTextRedactor.Redact` does not redact a credential written as a JSON property: `Redact("{\"password\":\"s3cr3t\"}")` returns the input verbatim (measured while finishing #1384).

## Why

`CredentialPairRegex` requires `\b(key)\b\s*[=:]` — in JSON the key's own closing quote sits between the key and the colon, so the pair never matches. A JSON body is exactly how a Creatio service answer, an environment-registration echo or a serialized DTO carries credentials, and `Redact` runs over already-serialized tool envelopes (`ClioRunTool.RedactFailureContent`).

## Proposed rule

- `"(password|pwd|pass|secret|token|api[_-]?key|client[_-]?secret|access[_-]?key|connection ?string|authorization|auth|bearer|cookie|…)"\s*:\s*(?:"[^"]*"|null|true|false|-?\d+(?:\.\d+)?)` → `"key":"[redacted]"`, so the surrounding payload stays valid JSON.
- Cover BOTH escaped forms of the quotes: `\"key\"` and `"key"` (the latter is what the default `System.Text.Json` encoder emits for a nested serialized body; an e2e test in #1384 proved the naive form misses it).
- Run it BEFORE `CredentialPairRegex`, otherwise that rule consumes the pair and leaves a dangling quote that breaks the JSON.
- Tests for both escaped forms, ordering vs `CredentialPairRegex`, and that a non-secret JSON property is untouched.

## Sequencing

Do this after #1473 (issue #1375) merges — the redactor moves to `clio/Common/SensitiveErrorTextRedactor.cs` there — and after #1493 (issue #1380, `EmailRegex`). The e2e-local post-pass added in #1384 (`clio.mcp.e2e/Support/Results/McpResultDiagnostics.cs`) becomes redundant once this lands and should be removed in the same PR.

Related: #1384, #1375, #1380, #1333.

### PR DESCRIPTION
**Pull request by Alexandr-Kravchuk**
🔗 **Issue:** [#1497](https://github.com/Advance-Technologies-Foundation/clio/issues/1497)

## What

`SensitiveErrorTextRedactor.Redact` now redacts a credential written as a JSON property — `"password":"s3cr3t"` — which `CredentialPairRegex` cannot reach at all: that rule needs `\b(key)\b\s*[=:]`, and the key's own closing quote sits between the key and the colon. Measured on the shipped rules, `Redact("{\"password\":\"s3cr3t\"}")` returned its input verbatim.

The whole change is one regex, one match evaluator and one key-set split:

- **One generated regex**, three quote spellings captured as `<q>` and written back unchanged: a literal `"`, the hand-written `\"`, and the `"` that `System.Text.Json`'s default encoder actually emits for a nested serialized body. The value is a JSON string in the **same** spelling, or `null` / `true` / `false` / a number. The replacement is `<q>key<q>:<q>[redacted]<q>`, so the document stays parseable.
- **The rule runs FIRST in the chain.** Measured, not reasoned: moving it after `BearerTokenRegex` or `CredentialPairRegex` turns both cases of `Redact_ShouldKeepEscapedJsonParseable_WhenTheValueAlsoMatchesALaterRule` red — their bare value class stops at a quote but takes the backslash before it, so on `\"authorization\":\"Bearer abc\"` and `\"cookie\":\"BPMCSRF=abc\"` the value's closing escape goes with it and a lone backslash is left behind.
- **The key words are declared once** and split into two sets. `uid` / `user id` / `data source` / `initial catalog` are deliberately kept out of the JSON set: `UId` is how Creatio spells the identifier property on every schema, package and descriptor payload.
- `CredentialPairRegex` and `BearerTokenRegex` keep their **base-branch value arms, byte for byte**.

## Simplified after review

The earlier revision of this PR carried a 177-line `JsonValueScanner`, object/array values, truncated/unterminated value arms, escaped-level content scanning, a 200-character key guard, modified `CredentialPairRegex` / `BearerTokenRegex` value arms and adversarial 100 KB tests — **+1422/−16 across 4 files**.

@kirillkrylov's review found no production path that puts credentials into an error message as JSON properties and classified those reproductions as helper-level, synthetic edge cases. We agree. Everything built for them is removed, including the `CredentialPairRegex` / `BearerTokenRegex` changes (and with them the P2 fix in `243a46d5e`) — **those arms are back to base behaviour**, so the truncated-quoted-credential case behaves exactly as it does on `master` today. `clio/Common/JsonValueScanner.cs` is deleted.

### Accepted limits

Not handled, knowingly. In every one of them the property is left **exactly as found**, so the document still parses:

- an object or array value — a regex cannot balance brackets;
- a value sliced by an input cap before its closing quote arrived;
- mixed quote spellings inside one document — the closing quote is a backreference to the opening one, and `System.Text.Json` emits one spelling consistently;
- an inner escaped quote inside an **escaped-level** value: it ends the match early and the tail is not redacted (`System.Text.Json` writes that inner quote as `\\` plus the spelling, and the escape unit consumes only the first two backslashes);
- a doubly-escaped document — an envelope serialized inside an envelope.

This redactor is a last line of defence, not the only one.

## Stacking

Base is `Alexandr-Kravchuk/issue-1380` (PR #1493), which itself sits on `Alexandr-Kravchuk/issue-1375` (PR #1473). **Retarget to `master` after #1493 merges.** The e2e-local post-pass added by PR #1498 (`clio.mcp.e2e/Support/Results/McpResultDiagnostics.cs`) becomes redundant once this lands; it is removed in a follow-up after #1498 merges, since that file is not on this branch.

## Verification

- Unit: `dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit"` — **12286 passed / 28 failed / 40 skipped**. The 28 failing names are identical to the macOS-only baseline measured on a clean `origin/Alexandr-Kravchuk/issue-1380` worktree in the same run (`diff` of the two sorted name lists is empty). Redactor fixture: **105/105**.
- Ordering: verified by temporarily moving the JSON rule after `BearerTokenRegex` and after `CredentialPairRegex` — both cases of the ordering test go red, then green again once reverted.
- Adversarial pass over the final diff (valid JSON in → invalid out; a `"` case that no longer redacts; regex timeout attributes): 16 shapes probed, including nested containers, a sliced value, a 200-backslash run, non-ASCII escapes, glued keys and `UId`. No valid JSON input produces invalid output. The one invalid-output shape found, `{"message":"password: \"abc\""}`, is **base `CredentialPairRegex` behaviour**, unchanged by this PR and per the review not a production path.
- Earlier stand run (kept for the record, on the larger revision): a local HTTP relay injected `{"password":"s3cr3t","server":"db.internal"}` into a DataService failure message against `ts1-core-dev04`. That proved the **MCP envelope path** — the redactor is reached and the envelope stays parseable — it did **not** demonstrate a production server path that writes a credential as a JSON property. No such path has been demonstrated by either side.

MCP reviewed, no update required. Docs reviewed, no update required (internal helper, no command surface). Knowledge record `docs/knowledge/Common/json-credential-rules-run-first-in-redact.md` rewritten to the kept facts.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Closes #1497

### CODE REVIEW

File: clio/Common/SensitiveErrorTextRedactor.cs [hunk 1]
```diff
@@ -75,15 +75,186 @@ internal static partial class SensitiveErrorTextRedactor {
...
+	// so `password: \"a\\\"b\"` leaves `b` behind. The document stays parseable; only the tail leaks.
+	private const string EscapedQuotedValuePattern =
+		@"(?<q>\\""|\\u0022)(?:[^""\\]|\\.)*?\k<q>";
+
+	// The same escaped value when the input was SLICED before its closing escape arrived. Without this arm
+	// the bare class below cannot start on it either - its lookahead refuses the opening escape - so the
+	// fragment would ship untouched. Non-capturing throughout: CredentialPairRegex reads its key from
+	// Groups[1], and an unnamed group added here would renumber it.
+	private const string UnterminatedEscapedQuotedValuePattern =
+		@"(?:\\""|\\u0022)(?:(?!\\""|\\u0022)[^\s,;""'])*";
```
Comment (kirillkrylov) [inline clio/Common/SensitiveErrorTextRedactor.cs]: **[P2] Consume the whole truncated quoted credential**

The new unterminated escaped-quoted arm stops at whitespace, commas, semicolons, and apostrophes even after it has consumed an opening quote. Against this PR's built assembly, the valid outer JSON `{"message":"password: \"first secretTail"}` becomes `{"message":"password=[redacted] secretTail"}`: the password tail remains in the MCP diagnostic. Serializer-generated probes reproduce the space/comma/semicolon case with both the default System.Text.Json encoder and UnsafeRelaxedJsonEscaping. This is incomplete protection in the newly added sliced-value handling, rather than a new leak relative to the base. Consume quoted content through the enclosing-string boundary/end of input instead of applying bare-value separators, and extend the sliced-value tests with these separators while asserting outer JSON parseability.

**Review by kirillkrylov — COMMENTED** [review]
Reviewed commit 57c133620c932868267eb73a3884f66b0e1d8605 across correctness, security, performance, maintainability, testing, intent alignment, and KISS.

Validation on Windows/.NET 10 in an isolated detached worktree:
- dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit&FullyQualifiedName~SensitiveErrorTextRedactor": 158 passed, 0 failed.
- dotnet test clio.tests/clio.tests.csproj --no-build --filter "Category=Unit": 12,453 passed, 25 skipped, 0 failed.
- Independent probe against the built clio.dll: 10/10 serializer-generated nested JSON cases preserved both outer/inner parseability, redacted credentials, and retained sibling values. Separate truncated-quoted-password probes confirmed the inline P2 finding with spaces, commas, and semicolons in both supported quote spellings.

One P2 finding: incomplete redaction of truncated quoted key=value credentials. This is a gap in the new handling, not a newly introduced leak relative to the base. The scanner's overall design is justified by fragments, escaping, and nested containers. No Creatio deployment was performed. Existing documented limitations (including deeper escaping) remain outside this validation's proof.

File: clio/Common/SensitiveErrorTextRedactor.cs (hunk 1 above)
Comment (Alexandr-Kravchuk) [reply clio/Common/SensitiveErrorTextRedactor.cs]: Confirmed and fixed in 243a46d5e — your reproduction is exactly right.

The arm was applying the BARE value separators after it had already consumed an opening escape. Once that escape is consumed the cursor is inside the enclosing JSON string, where a space, comma, semicolon and apostrophe are ordinary content rather than value boundaries, so the match ended early and the tail shipped. It now runs to the enclosing string's own closing quote — a bare `"`, the one character that cannot appear inside it unescaped — or to the end of a sliced input. That keeps the envelope parseable, which was the reason the bare class stops at a quote in the first place.

Added five checks as you suggested: one per separator (space, comma, semicolon, apostrophe), each built with `JsonSerializer.Serialize` and re-parsed with `JsonDocument.Parse` after redaction, plus one pinning that the match does not run past the enclosing string so a following property survives. The four separator cases fail on the previous pattern, so they are not vacuous. 163 redactor tests pass.

**Review by kirillkrylov — APPROVED** [review]
@Alexandr-Kravchuk — green from our side. You can merge this PR if you still consider the change necessary. We are not asking for further work on the synthetic edge case, and it is not a merge blocker for us.

Follow-up validation report:

- Provisioned a dedicated Creatio 10.1.585 / .NET 8 / PostgreSQL environment. Created a user through the native AdministrationService, set a password containing the relevant quote/space pattern, and successfully authenticated with it.
- Exercised actual MCP reads and invalid writes using the reviewed build. The tested server-side failures did not expose the password tail. An OData password-field attempt was rejected locally against metadata, so we do not count that as server-side proof.
- A deliberately invalid native administration request did produce a PostgreSQL error containing the test password. Passing that captured real response through the PR redactor removed it completely, both as decoded prose and serialized JSON, preserving JSON parseability.
- Backend-source inspection found that password-policy errors format policy requirements rather than the submitted password, and the inspected administration/DataService handlers forward complete exception messages. We did not establish a normal native path producing the truncated input behind my earlier P2 comment.

The earlier reproduction demonstrates a helper-level defensive edge case, not a demonstrated production credential leak. My initial feedback should have made that practical distinction clearer.

Evidence boundary: local tests and live checks used 57c13362 (158 redactor tests passed; 12,453 unit tests passed, 25 skipped; 10 nested JSON roundtrips passed). I have now inspected the follow-up fix and regression tests in 243a46d5e; current CI is green. I did not rerun the local/live matrix on that newer commit. The original review thread is resolved, and I have no remaining merge-blocking findings. The disposable test user has been disabled.

**Comment by Alexandr-Kravchuk** [comment]
@kirillkrylov — thank you for the follow-up validation, and especially for provisioning a real stand to check it. We agree with your conclusion: the reproductions behind the larger version of this change were helper-level and synthetic, not a demonstrated production leak.

So the PR has been reduced to the measured facts only, and force-pushed as one clean commit (`911d39b`):

- **What stays** — one regex for a credential written as a JSON property, in the three quote spellings (including the `"` one `System.Text.Json` actually emits, which is why the plain-quote-only rule would have caught nothing on a real envelope); the rule running **first** in the chain (measured — move it after `BearerTokenRegex` or `CredentialPairRegex` and the ordering test goes red both ways); and the key-set split that keeps `uid` out of the JSON form so Creatio's `UId` property is not blanked.
- **What is gone** — `JsonValueScanner.cs` (all 177 lines), object/array values, the truncated/unterminated value arms, escaped-level content scanning, the 200-character key guard and the adversarial 100 KB tests.
- **Including your P2** — the `CredentialPairRegex` and `BearerTokenRegex` value arms are reverted to the base branch byte for byte, which removes the code your P2 was about along with the fix in `243a46d5e`. Those two rules therefore behave on a truncated quoted credential exactly as they do on `master` today; no change, no regression, no new machinery.

The diff went from **+1422/−16 across 4 files** to **+405/−3 across 3 files**, and the remaining limits (containers, sliced values, mixed spellings, deeper escaping) are now stated plainly as accepted — in each of them the property is left exactly as found, so the document still parses.

Verification on the reduced diff: redactor fixture 105/105; full `Category=Unit` 12286 passed / 28 failed / 40 skipped, with the 28 names identical to the base branch's macOS-only baseline measured in the same session. An adversarial pass over 16 JSON shapes found no valid-JSON input that comes out invalid.

Could you take one more look at the reduced diff and re-approve if it reads right to you? Much smaller read this time.

**Review by kirillkrylov — COMMENTED** [review]
Reviewed current commit 911d39b783a3121091ac0c473c2babe0919b0d0f, including independent intent/KISS, correctness/security, quality/performance and test reviews. The reduced approach fits the intended bounded JSON-property redaction change. No material defect attributable to this change was found in the executed scope.

Real Creatio validation: YES. Built this exact head and exercised it against a newly provisioned exclusive disposable instance, pr1504-reduced-0912 (HTTPS port 40023), Creatio 10.1.585.0 / .NET 8 / PostgreSQL. Native AdministrationService created a real user, changed its password, and authentication with that new password succeeded. Actual stdio MCP dispatch performed account reads, a Contact update with persisted readback, and invalid-value/missing-schema failure paths. Real Creatio-generated ItemNotFoundException diagnostics containing deliberately prepared JSON-shaped schema names had password, authorization and cookie values redacted; UId and surrounding stage text survived. All returned MCP envelopes parsed successfully. These are actual server responses, without a relay or injected response. The prepared inputs prove redaction on a real error path; they do not establish that normal Creatio failures expose stored credentials.

The account was disabled through AdministrationService and Active=false was verified through MCP. The disposable instance is retained; temporary test-password files were removed.

Synthetic/unit validation: 105 redactor tests passed. Full Category=Unit run: 12,400 passed, 25 skipped, 0 failed. Separate StageEventLivenessHeartbeatTests rerun: 8 passed. Commands: `dotnet test clio.tests/clio.tests.csproj --filter 'Category=Unit&FullyQualifiedName~SensitiveErrorTextRedactor'`; `dotnet test clio.tests/clio.tests.csproj --no-build --filter 'Category=Unit'`; `dotnet test clio.tests/clio.tests.csproj --no-build --filter 'FullyQualifiedName~StageEventLivenessHeartbeatTests'`.

Runtime-reproduced issues: None found within these selected scenarios.
Synthetic-test findings: No material new defect retained. Unsupported/truncated/deeper escaping shapes are outside the real-environment validation scope.
Static findings: No material blocking finding in this diff.

CI blocker: GitHub Actions unit-4 failed one pre-existing test, RunStage_ShouldReAnnounceTheRunningStage_WhenItsActionOutlastsTheRefreshInterval (StageEventLivenessHeartbeatTests.cs:73): expected more than one running event but got one. Unit Tests is the aggregate gate reflecting that same failure, not a second failing test. The test sleeps for 700 ms while the heartbeat uses Task.Run plus a 100 ms delay; scheduler delay is a plausible cause, not proven by the log. Both this test and StageEventEmitter are unchanged in this PR and passed locally. Failure log: https://github.com/Advance-Technologies-Foundation/clio/actions/runs/34689282000/job/103541648239

Alex: the scoped code/runtime review is green, but the current required CI check is red, so this is a COMMENT rather than renewed approval. Please rerun/investigate that heartbeat failure before merging. No general production credential leak was established by this validation.

**Comment by Alexandr-Kravchuk** [comment]
@kirillkrylov — thank you, you called the CI blocker exactly right, and it turned out to have a concrete root cause rather than being pure scheduler luck.

`RunStage_ShouldReAnnounceTheRunningStage_WhenItsActionOutlastsTheRefreshInterval` failed three consecutive reruns on this head, so I stopped rerunning and looked for the reason. **Master had already fixed this test** in `b0b1d816` ("Merge master and synchronize heartbeat tests with observed refresh", Sep 11): the fixed 700 ms `Thread.Sleep` was replaced with a bounded `SpinWait.SpinUntil(..., 10s)` that waits for a real refresh instead of racing the scheduler. This PR sits on a stack whose bottom (#1473) was 140 commits behind master, so it was still running the old, flaky version — which is why master is green and this branch was not.

Fix applied: merged `origin/master` through the whole stack, bottom-up, with no conflicts anywhere:

- #1473 `issue-1375` ← `master` (`1917a3f1`)
- #1493 `issue-1380` ← `issue-1375` (`621282f3`)
- #1504 `issue-1497` ← `issue-1380` (`16e51be8`)

The heartbeat test and `StageEventEmitter` are now byte-identical to master on this branch (verified by diff), and no change of mine touches either. Local verification on the merged head: `Category=Unit&(FullyQualifiedName~StageEventLiveness|FullyQualifiedName~SensitiveErrorTextRedactor)` — 113 passed, 0 failed.

The redaction change itself is untouched by this merge. Waiting on CI now, and I will re-request your review once it is green.
